File: apps/backend/src/database/models/cluster.model.ts
import {
  Table,
  Column,
  Model,
  DataType,
  DeletedAt,
  UpdatedAt,
  CreatedAt,
  Default,
  PrimaryKey,
  HasOne,
  ForeignKey,
  HasMany,
  BelongsTo,
} from 'sequelize-typescript';
import { ICluster } from '@shared/shared/src/types/cluster';
import { CreationOptional, DataTypes } from 'sequelize';
import { FundingApplication } from './funding_application';
import { Duration, SyncStatus } from '@shared/shared/src/enums';
import { Collection } from './collection.model';

@Table({ tableName: 'cluster' })
export class Cluster extends Model<ICluster> implements ICluster {
  @PrimaryKey
  @Default(DataTypes.UUIDV4)
  @Column(DataTypes.UUID)
  declare id: CreationOptional<string>;

  @Column(DataType.STRING)
  declare name: string;

  @Column(DataType.STRING)
  declare status: string;

  @Column(DataType.STRING)
  declare about: string;

  @Column(DataType.STRING)
  declare duration: Duration;

  @Column(DataType.BIGINT)
  declare roi: number;

  @Column(DataType.BIGINT)
  declare total_funds_raised: number;

  @Column(DataType.BIGINT)
  declare target_fundraising_amount: number;

  @Column(DataType.STRING)
  declare repayment: string;

  @Column(DataType.BOOLEAN)
  declare is_active: boolean;

  @Column(DataType.STRING)
  declare cover_image: string;

  @Column(DataType.STRING)
  declare start_date: string;

  @Column(DataType.STRING)
  declare end_date: string;

  @Column(DataType.STRING)
  declare description: string;

  @Column(DataType.DATE)
  declare sync_date: string;

  @Column(DataType.STRING)
  declare sync_status: SyncStatus;


  @HasOne(() => FundingApplication)
  declare application: FundingApplication;

  @BelongsTo(() => Collection)
  declare collection: Collection;

  @ForeignKey(() => Collection)
  @Column(DataType.UUID)
  declare collection_id: string;

  @CreatedAt
  declare createdAt: string;

  @UpdatedAt
  declare updatedAt: string;

  @DeletedAt
  declare deletedAt?: string;
}

File: apps/spine/lib/drift/model/cluster.model.dart
import 'package:drift/drift.dart';
import 'package:spine/drift/model/base.dart';
import 'package:spine/drift/model/collection.model.dart';


class Cluster extends Table with TableMixin {
    late final name = text()();
    late final status = text()();
    late final about = text()();
    late final duration = integer()();
    late final roi = integer()();
    late final totalFundsRaised = integer()();
    late final targetFundRaisingAmount = integer()();
    late final repayment = text()();
    late final isActive = boolean()();
    late final coverImage = text()();
    late final startDate = dateTime()();
    late final endDate = dateTime()();
    late final description = text()();

    late final collectionId = text().nullable().references(Collection, #id)();
}
